third_party/nvidia/hopper/lib/Transforms/WarpSpecialization/TMEMAlloc1D.cpp

```diff
@@ -0,0 +1,229 @@
+    // the example.
+    // TODO: Should we copy over loop information?
+    auto partitionAttr = oldOp->getAttr("ttg.partition");
+    newOp->setAttr("ttg.partition", partitionAttr);
+  }
+
+  void setExpandedInput(tt::ExpandDimsOp expandedInput) {
+    this->expandedInput = expandedInput;
+  }
+
+  tt::ExpandDimsOp getExpandedInput() {
+    assert(expandedInput != nullptr && "Must call setExpandedInput");
+    return expandedInput;
+  }
+
+  void setAllocOp(ttng::TMEMAllocOp allocOp) { this->allocOp = allocOp; }
+
+  ttng::TMEMAllocOp getAllocOp() {
+    assert(allocOp != nullptr && "Must call getAllocOp()");
+    return allocOp;
+  }
+
+  RankedTensorType getResultTensorType(Operation *op, size_t expectedSize) {
+    auto outputType = dyn_cast<RankedTensorType>(op->getResult(0).getType());
+    if (!outputType || outputType.getShape().size() != 2) {
+      assert("Invalid tensor input");
+    }
+    return outputType;
+  }
+
+  ttng::TMEMAllocOp alloc1DTMEMBuffer() {
+    auto expandedInput = getExpandedInput();
+    auto oldRetType = getResultTensorType(expandedInput, 2);
+    SmallVector<int64_t> shape = {oldRetType.getShape().begin(),
+                                  oldRetType.getShape().end()};
+    auto context = builder.getContext();
+    auto oldEncoding = oldRetType.getEncoding();
+    Attribute tensorMemorySpace =
+        ttng::TensorMemorySpaceAttr::get(builder.getContext());
+    // TODO(njriasan): Do we need to handle the ScaleDotElemType::E2M1 && transA
+    // case at all from TCGen5MMAScaledOp::getBlockM?
+    auto blockM = shape[0];
+    auto elemType = oldRetType.getElementType();
+    unsigned elemBitWidth = elemType.getIntOrFloatBitWidth();
+    assert((elemBitWidth == 16 || elemBitWidth == 32) &&
+           "TMEM Layout don't support fp8");
+    bool unpacked = elemBitWidth != 16;
+    ArrayRef<unsigned> CTASplitNum =
+        ttg::getCTALayout(oldEncoding).getCTASplitNum();
+    auto encoding = ttng::TensorMemoryEncodingAttr::get(
+        builder.getContext(), blockM, shape[1],
+        /*unpacked=*/unpacked, CTASplitNum[0], CTASplitNum[1]);
+    auto tmemDesc =
+        ttg::MemDescType::get(shape, elemType, encoding, tensorMemorySpace,
+                              /*mutableMemory=*/true);
+
+    builder.setInsertionPointToStart(globalStart);
+    auto allocCall = builder.create<ttng::TMEMAllocOp>(
+        expandedInput->getLoc(), tmemDesc,
+        builder.getType<ttg::AsyncTokenType>(),
+        /*src=*/Value());
+    return allocCall;
+  }
+
+  void TMEMStore1D(Operation *producer,
+                   std::optional<ttng::TMEMAllocOp> allocOpBuffer) {
+    // Expand from 1D -> 2D
+    auto oldRetType = getResultTensorType(producer, 1);
+    builder.setInsertionPointAfter(producer);
+    auto expandDims = builder.create<tt::ExpandDimsOp>(
+        producer->getLoc(), producer->getResult(0), 1);
+    copyAttrs(producer, expandDims);
+    setExpandedInput(expandDims);
+    ttng::TMEMAllocOp allocOp;
```

Comment:
Why have a local variable shadowing the class member here?

Reply:
Resolved. Good catch!